### PR TITLE
Add support for Git Log

### PR DESCRIPTION
**Pull request by i053322**
What does this PR do?

What issues does this PR fix or reference?

Previous Behavior

Remove this section if not relevant

New Behavior

This content may also be included in the release notes.

Tests written?

Yes/No

Docs requirements?

Include the content for all the docs changes required. 
1. API changes

2. User docs changes

Please review Che's Contributing Guide for best practices.

Add support for skip and maxCount when calling the LOG API.
Add support to get the log for a specific folder or file in the repository.
For every commit that is returned when calling the LOG API, the following data was added:
The author of the commit
A list of branch names that the commit is related to.
A list of file names that were changed/deleted/added in the commit.
A list of commit parents (to enable for example clients to paint a graph)

Signed-off-by: yossi balan <yossi.balan@sap.com>

Change-Id: I9c66b71b13164b2182c614077d0aba3b68b43235

### CODE REVIEW

**Comment by i053322** [comment]
@skabashnyuk @vinokurig@ShimonBenYair

**Comment by i053322** [comment]
I added the implementation from eclipse/che https://github.com/eclipse/che/pull/2046

File: platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/ErrorCodes.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    public static final int NO_COMMITTER_NAME_OR_EMAIL_DEFINED = 15216;
+    public static final int UNABLE_GET_PRIVATE_SSH_KEY         = 32068;
+    public static final int UNAUTHORIZED_GIT_OPERATION         = 32080;
+    public static final int UNAUTHORIZED_SVN_OPERATION         = 32090;
+    public static final int MERGE_CONFLICT                     = 32062;
+    public static final int FAILED_CHECKOUT                    = 32063;
+    public static final int FAILED_CHECKOUT_WITH_START_POINT   = 32064;
+    public static final int INIT_COMMIT_WAS_NOT_PERFORMED      = 32082;
+}
```
Comment (vinokurig) [inline platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/ErrorCodes.java]: Please add new line

File: platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/rest/shared/dto/ExtendedError.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+     *         error attributes
+     */
+    void setAttributes(Map<String, String> attributes);
+
+    ExtendedError withAttributes(Map<String, String> attributes);
+
+    @Override
+    ExtendedError withMessage(String message);
+
+}
```
Comment (vinokurig) [inline platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/rest/shared/dto/ExtendedError.java]: Please add new line

File: platform-api/che-core-api-git/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Create a {@link DiffCommitFile} object based on a given file new location.
+     *
+     * @param newPath
+     *         file new location
+     * @return a {@link DiffCommitFile} object that contains information about changes of the committed file
+     */
+    DiffCommitFile withNewPath(String newPath);
+}
```
Comment (vinokurig) [inline platform-api/che-core-api-git/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java]: Please add new line

File: platform-api/che-core-api-git/src/main/java/org/eclipse/che/api/git/shared/LogRequest.java [hunk 4]
```diff
@@ -19,14 +21,90 @@
...
+    void setFileFilter(List<String> fileFilter);
+
+    /**
+     * Create a {@link LogRequest} object based on a given range of files
+     *
+     * @param fileFilter
+     *         range of files
+     */
+    LogRequest withFileFilter(List<String> fileFilter);
 }
```
Comment (vinokurig) [inline platform-api/che-core-api-git/src/main/java/org/eclipse/che/api/git/shared/LogRequest.java]: Please add new line

File: platform-api/che-core-api-git/src/main/java/org/eclipse/che/api/git/shared/Revision.java [hunk 5]
```diff
@@ -19,43 +20,145 @@
...
+
+    /**
+     * Create a {@link Revision} object based on a given list of commit parents
+     *
+     * @param commitParents
+     *         a list of commit parents
+     * @return a {@link Revision} object that contains information about revision
+     */
+    Revision withCommitParent(List<String> commitParents);
 }
```
Comment (vinokurig) [inline platform-api/che-core-api-git/src/main/java/org/eclipse/che/api/git/shared/Revision.java]: Please add new line

File: platform-api/che-core-api-git/src/test/java/org/eclipse/che/git/impl/LogTest.java [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+
+        assertEquals(secondAndThirdAndFourthCommits.get(0).getMessage(), "add 4.txt file");
+        assertEquals(secondAndThirdAndFourthCommits.get(1).getMessage(), "add 3.txt file");
+        assertEquals(secondAndThirdAndFourthCommits.get(2).getMessage(), "add 2.txt file");
+
+        assertEquals(thirdAndFourthCommits.get(0).getMessage(), "add 4.txt file");
+        assertEquals(thirdAndFourthCommits.get(1).getMessage(), "add 3.txt file");
+    }
+}
```
Comment (vinokurig) [inline platform-api/che-core-api-git/src/test/java/org/eclipse/che/git/impl/LogTest.java]: Please add new line

**Review by vinokurig — APPROVED** [review]

**Comment by i053322** [comment]
fixed

**Comment by i053322** [comment]
@vinokurig  - do you have more comments ?

**Comment by vinokurig** [comment]
@tareqhs could you merge it please if it is OK for you?

File: che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 7]
```diff
@@ -593,39 +607,133 @@ public void init(InitRequest request) throws GitException {
...
                 commits.add(revision);
             }
             return new LogPage(commits);
         } catch (GitAPIException | IOException exception) {
-            throw new GitException(exception.getMessage(), exception);
+            String errorMessage = exception.getMessage();
+            if (ERROR_LOG_NO_HEAD_EXISTS.equals(errorMessage)) {
+                throw new GitException(errorMessage, ErrorCodes.INIT_COMMIT_WAS_NOT_PERFORMED);
+            } else {
+                LOG.error("Failed to retrieve log. ", exception);
```
Comment (tareksha) [inline che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java:638]: the exception message should be included in the log message

**Pull request merged**